crates/neon/src/types_impl/mod.rs

```diff
@@ -362,6 +424,54 @@ impl JsString {
         }
     }
 
+    /// Convert the JavaScript String into a UTF-16 encoded [`Vec<u16>`].
+    ///
+    /// The returned vector is guaranteed to be valid UTF-16. Therefore, any external crate that
+    /// handles UTF-16 encoded strings, can assume the content to be valid and skip eventual
+    /// validation steps.
+    ///
+    /// # Example
+    ///
+    /// A function that expects a single JavaScript string as argument and prints it out as a raw
+    /// vector of `u16`s.
+    ///
+    /// ```rust
+    /// # use neon::prelude::*;
+    /// fn print_string_as_utf16(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    ///     let s = cx.argument::<JsString>(0)?.to_utf16(&mut cx);
+    ///     println!("JavaScript string as raw UTF-16: {:?}", s);
+    ///
+    ///     Ok(cx.undefined())
+    /// }
+    /// ```
+    ///
+    /// Again a function that expects a single JavaScript string as argument, but utilizes the
+    /// [`widestring`](https://crates.io/crates/widestring) crate to handle the raw [`Vec<u16>`] as
+    /// a typical string.
+    ///
+    /// ```rust
+    /// # use neon::prelude::*;
+    /// fn print_with_widestring(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    ///     let s = cx.argument::<JsString>(0)?.to_utf16(&mut cx);
+    ///     let s = unsafe { widestring::Utf16String::from_vec_unchecked(s) };
+    ///     println!("JavaScript string as UTF-16: {}", s);
+    ///
+    ///     Ok(cx.undefined())
+    /// }
+    /// ```
+    pub fn to_utf16<'a, C: Context<'a>>(&self, cx: &mut C) -> Vec<u16> {
+        let env = cx.env().to_raw();
+
+        unsafe {
+            let capacity = sys::string::utf16_len(env, self.to_raw()) + 1;
+            let mut buffer: Vec<u16> = Vec::with_capacity(capacity as usize);
+            let p = buffer.as_mut_ptr();
+            std::mem::forget(buffer);
```

Comment:
I see this is how the existing function is written and appreciate keeping things consistent. However, when reviewing this, I noticed that this could result in a memory leak if `data_utf16` panics due a non-`Ok` response.

The nice side is this protects against UB if Node-API continues to write to the data after the panic, but that seems a bit extreme. This _seems_ like it could be simplified to using `unsafe fn set_len`.

@dherman what do you think? Is this safety a bit too extreme?

Reply:
Is the `std::mem::forget` actually needed at all?

There is probably some specific case that I'm missing here. Maybe related to panics as well?

I personally would have just passed the pointer `p` to `sys::string::data`/`sys::string::data_utf16`, without calling anything in between, then calling `set_len` as you mentioned, and last depending on the function either:
- `value`: use `String::from_utf8_unchecked` to turn the `Vec<u8>` into a `String` afterwards.
- `to_utf16`: Nothing, because it's already the expected `Vec<u16>`.

```rust
// in `value`
unsafe {
    let capacity = sys::string::utf8_len(env, self.to_raw()) + 1;
    let mut buffer: Vec<u8> = Vec::with_capacity(capacity as usize);
    let len = sys::string::data(env, buffer.as_mut_ptr(), capacity, self.to_raw());
    buffer.set_len(len as usize);
    String::from_utf8_unchecked(buffer) // turn into a String
}

// in `to_utf16`
unsafe {
    let capacity = sys::string::utf8_len(env, self.to_raw()) + 1;
    let mut buffer: Vec<u8> = Vec::with_capacity(capacity as usize);
    let len = sys::string::data(env, buffer.as_mut_ptr(), capacity, self.to_raw());
    buffer.set_len(len as usize);
    buffer // Just return the Vec
}
```